File: tests/src/robot.test.ts
import { Robot } from "../../src/robot.ts";

import { assertEquals } from "../../tests/deps.ts";

Deno.test("robot", async (t) => {
  await t.step("process to [0,0]", () => {
    const commands = "N,E,S,W".split(",");
    const robot = new Robot();
    robot.process(commands);
    assertEquals(robot.position, [0, 0]);
  });

  await t.step("process to [4,4]", () => {
    const commands = "N,E,N,E,N,E,N,E".split(",");
    const robot = new Robot(0, 0);
    robot.process(commands);
    assertEquals(robot.position, [4, 4]);
  });

  await t.step("process to [9,9]", () => {
    const commands = [];
    for (let i = 0; i < 9; i++) {
      commands.push("N");
      commands.push("E");
    }
    const robot = new Robot(0, 0);
    robot.process(commands);
    assertEquals(robot.x, 9);
    assertEquals(robot.y, 9);
    assertEquals(robot.position, [9, 9]);
  });

  await t.step("process to [0,0] from [9,9]", () => {
    const commands = [];
    for (let i = 0; i < 9; i++) {
      commands.push("S");
      commands.push("W");
    }
    const robot = new Robot(9, 9);
    robot.process(commands);
    assertEquals(robot.x, 0);
    assertEquals(robot.y, 0);
    assertEquals(robot.position, [0, 0]);
  });

  await t.step("process does nothing when given unknwon commands", () => {
    const commands = "X,Y,Z".split(",");
    const robot = new Robot();
    robot.process(commands);
    assertEquals(robot.position, [0, 0]);
  });
});

File: tests/src/robot_test.rb
# typed: true
require 'minitest/autorun'

require_relative '../../src/robot'

class RobotTest < Minitest::Test
  def test_process_to_origin
    commands = %w[N E S W]
    robot = Robot.new
    robot.process(commands)
    assert_equal [0, 0], robot.position
  end

  def test_process_to_coordinates
    commands = %w[N E N E N E N E]
    robot = Robot.new(0, 0)
    robot.process(commands)
    assert_equal [4, 4], robot.position
  end

  def test_process_to_edge
    commands = []
    9.times do
      commands.push('N', 'E')
    end
    robot = Robot.new(0, 0)
    robot.process(commands)
    assert_equal 9, robot.x
    assert_equal 9, robot.y
    assert_equal [9, 9], robot.position
  end

  def test_process_back_to_origin
    commands = []
    9.times do
      commands.push('S', 'W')
    end
    robot = Robot.new(9, 9)
    robot.process(commands)
    assert_equal 0, robot.x
    assert_equal 0, robot.y
    assert_equal [0, 0], robot.position
  end

  def test_process_unknown_commands
    commands = %w[X Y Z]
    robot = Robot.new
    robot.process(commands)
    assert_equal [0, 0], robot.position
  end
end
